src/misc/rwqueue.cpp

```diff
@@ -0,0 +1,96 @@
+/*
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *
+ *  Copyright (C) 2021-2021  The DOSBox Staging Team
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "rwqueue.h"
+
+#include <cassert>
+
+template <typename T>
+RWQueue<T>::RWQueue(size_t queue_capacity) : capacity(queue_capacity)
+{
+	assert(capacity > 0);
+}
+
+template <typename T>
+size_t RWQueue<T>::Size()
+{
+	std::lock_guard<std::mutex> lock(mutex);
+	return queue.size();
+}
+
+template <typename T>
+size_t RWQueue<T>::MaxCapacity() const
+{
+	return capacity;
+}
+
+template <typename T>
+bool RWQueue<T>::IsEmpty()
+{
+	std::lock_guard<std::mutex> lock(mutex);
+	return !queue.size();
+}
+
+template <typename T>
+void RWQueue<T>::CopyIn(const T &item)
+{
+	// wait until the queue has room to accept the item
+	std::unique_lock<std::mutex> lock(mutex);
+	while (queue.size() >= capacity)
+		has_room.wait(lock);
+
+	// add it, and notify the next waiting thread that we've got an item
+	queue.emplace(queue.end(), item);
+	lock.unlock();
+	has_items.notify_one();
+}
+
+template <typename T>
+void RWQueue<T>::MoveIn(T &item)
```

Comment:
Method name is in conflict with interface. This probably should be `::MoveIn(T &&item)`.

Reply:
`MoveIn(T &item)` performed the moving inside the call, so the caller didn't have to.  For example, calling `q.MoveIn(my_vector)` already tells me my object `myVector` has been 'moved'. 

If we use `T &&item`, then the caller needs to do the moving in the call `q.MoveIn(std::move(my_vector))`.
That's harder on my eyes, and it's redundant given we have two "move" words.

If we want to go this route, then we should get rid of the explicit `CopyIn` and `MoveIn` APIs, and use just one `Enqueue` similar to moodycamel, and then vary the in-bound type:

```
::Enqueue(const T& item) <-- copy
::Enqueue(T && item)  <-- move
```

This is similar to the STL too, so I guess that's a positive.

My only concern here is that these feel ambiguous to the user.  Stack Exchange is packed with people trying to figure out if the really need to use `std::move` or if the move is happening automatically.

There are also lots of experts who saying that explicitly calling `std::move` on STL objects is even more wasteful than not - and instead recommend writing interfaces to avoid it and let the compiler to pure rvalue substitution (which is even more efficient than `std::move()`).  I realize that's not the argument here - but I'm just mentioning it because it falls into this ambiguous trap.

Given we control the queue code itself, it will be easy for us to know what's happening - so I'm more Ok with something like `Enqueue(..)`.

Curious what you think.

Reply:
Edit: agreed, and renamed calls to:
 Enqueue(..)
 Dequeue()